### PR TITLE
Improve boot experience and help messages

### PR DESCRIPTION
**Pull request by 4086606**
* Don't show the Matlab terminal until it's reached the starting message.
* Base the starting message on the official MATLAB one.
* Drop "REPL" from the terminal name, the other language terminals don't suffix the lang terminal.

```
MATLAB Interactive Terminal (R2021a)

To get started, type one of these commands:
   helpwin           Provide access to help comments for all functions
   helpdesk          Open help browser
   demo              Access product examples in Help browser

For product information, visit https://www.mathworks.com.
```

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -2,9 +2,9 @@
 
 // The module 'vscode' contains the VS Code extensibility API
 // Import the module and reference it with the alias vscode in your code below
-import * as vscode from 'vscode';
-import * as path from 'path';
-import * as util from 'util';
+import * as vscode from "vscode";
+import * as path from "path";
+import * as util from "util";
 import * as fs from 'fs';
```
Comment (apommel) [inline src/extension.ts:8]: You may want to change the quotation marks to import fs too for consistency.

**Comment by apommel** [comment]
Thanks for your contribution! I'm okay with the changes overall. However, there are some issues when I test your branch on my end.

Capture

As you can see, the stdout redirection does not work for some reason, and the string formatting doesn't either.

Also you mention
> Don't show the Matlab terminal until it's reached the starting message.
But on my side it is still displayed as soon as the command is launched.

So overall, I like the consistencies change, but the new features/fix don't work for me.

**Comment by 4086606** [comment]
~~Kurwa MATLAB Engine~~ Oh crap, I know what's happening. MATLAB implicitly `print`s all unsuppressed variables.

I'll look into hiding the terminal from the user. Fwiw from `vscode-python` this seems broken in the API itself..

**Comment by 4086606** [comment]
Python doesn't have the issue, I'll take a look at their repo and see how they booted the terminal..

**Comment by apommel** [comment]
To be honest, I had a look at not showing the terminal directly before, but I think the only way to get this level of control is to implement a pseudo-terminal. However, this implies redoing a lot of the code. I actually tried it because I thought it would allow me to solve other issues, but I had trouble making it feature-equal with what exists currently, and did not solve all of the problems I expected to solve as easily as I thought.

File: interfaces/standard/matlab_interface.py [hunk 2]
```diff
@@ -9,13 +9,16 @@
     pass
 
 import os
+from io import StringIO
+from contextlib import redirect_stdout
```
Comment (apommel) [inline interfaces/standard/matlab_interface.py:13]: It is not necessary to import `redirect_stdout` anymore.

Comment (4086606) [reply interfaces/standard/matlab_interface.py]: Nice catch @apommel resolved in 1cd12b6.

File: src/extension.ts (hunk 1 above)
Comment (4086606) [reply src/extension.ts]: Resolved in 556208e!

File: interfaces/standard/matlab_interface.py [hunk 3]
```diff
@@ -29,70 +30,115 @@ def __init__(self):
...
+                    demo             Access product examples in Help browser
+                
+                For product information, visit https://www.mathworks.com.\
+                '''.format(release=self.release())
+                print(dedent(intro))
+
+            except Exception as e:
+                self.clear()
+                print("MATLAB Engine for Python exited prematurely.")
+                print(str(e))
```
Comment (apommel) [inline interfaces/standard/matlab_interface.py:54]: The exception can be any Matlab exception here, not only a crash from the Matlab Engine

File: interfaces/standard/matlab_interface.py [hunk 4]
```diff
@@ -29,70 +30,115 @@ def __init__(self):
...
+
         else:
-            print("Could not start Matlab")
+            print("Launching MATLAB failed: Error starting MATLAB process in MATLAB Engine for Python.")
 
     def clear(self):
         os.system(self.cls_str)
 
+    def release(self):
+        release_str = "version '-release';"
```
Comment (apommel) [inline interfaces/standard/matlab_interface.py:63]:
```suggestion
        release_str = "version('-release');"
```
This instruction fails. When trying directly on Matlab, I see that it fails when trying to assign it to a variable (ie `a = version '-release';"). The statement alone works, and assigns the value to ans correctly. Weird. But using `version('-release');' seems to work :)

File: interfaces/standard/matlab_interface.py [hunk 5]
```diff
@@ -29,70 +30,115 @@ def __init__(self):
...
-                print("Running: \"{}\"".format(script_path))
+                print("File: \"%s\""%script_path)
                 self.eng.run(script_path, nargout=0)
+
             except MatlabTerminated:
-                print("Matlab terminated. Restarting the engine...")
+                print("MATLAB process was terminated. Restarting the engine...")
                 self.eng = matlab.engine.start_matlab()
-                print("Matlab restarted")
+                print("Matlab process restarted.")
```
Comment (apommel) [inline interfaces/standard/matlab_interface.py:76]: You capitalized MATLAB almost everywhere but not here

File: interfaces/standard/matlab_interface.py (hunk 3 above)
Comment (4086606) [reply interfaces/standard/matlab_interface.py:53]: Resolved in 63171c8 - "MATLAB Engine for Python is missing" can get reported now 😄

File: interfaces/standard/matlab_interface.py (hunk 5 above)
Comment (4086606) [reply interfaces/standard/matlab_interface.py]: Resolved in 088ff6d 🚀

**Comment by 4086606** [comment]
@apommel I solved the terminal issue in f9013e8. Turns out vscode-python uses `createTerminal` without passing text. Switching from `Terminal.sendText` to the `TerminalOptions` keys fixed it!

**Comment by apommel** [comment]
You did quite a change in the way the Python executable is looked for, and I am not sure what everything is supposed to do. However, when leaving the `pythonPath` option blank, this ends up looking for an executable called `undefinedpython.exe` and obviously fails.

**Comment by 4086606** [comment]
It's definitely working on my end. The extension looks for Python 3.8, 3.7 or 2.7 inside the `PATH` environment variable. If it's not available in the PATH variable, it will suddenly and aggressively break.

**Comment by apommel** [comment]
Yeah I get what you are doing. It does not work on my end because I installed it though miniconda, so it is not inside a Python38, Python37 or Python28 folder.

**Comment by 4086606** [comment]
Maybe this is a case of overengineering. A simple `python` should work, then the README gets updated to clarify that the 3.7 or 3.8 Python installation must be the first Python executable available in the PATH.

E.g. my PATH looks like this:

```
C:\Program Files\Python38\Scripts\;C:\Program Files\Python38\;C:\Program Files\Python39\Scripts\;C:\Program Files\Python39\;C:\Program Files (x86)\Common Files\Oracle\Java\javapath;C:\Windows\System32;...
```

**Comment by apommel** [comment]
Yeah otherwise it is possible to use the `pythonPath` option, which is why I implemented it in the first place. The good thing about your way is that it would have been possible to catch if Python was missing from `PATH`, but I think it's enough to catch it when trying to run `check_dependencies.py`.

**Comment by 4086606** [comment]
Right that's our work done! Thanks for your patience and all the reviews

PRs for upgrading TSLint and adding a terminal profile coming soon™️

File: src/extension.ts [hunk 6]
```diff
@@ -2,108 +2,121 @@
...
 		python_path = getPythonPath();
 		correct_setup = checkSetup();
 		if (correct_setup) {
-			const terminal = vscode.window.createTerminal(terminalLaunchOpt);
-			terminal.sendText(python_path + util.format(" \"%s\"", script_path));
-			terminal.show(false);
+			const opts: vscode.TerminalOptions = { ...terminalLaunchOptions };
+			opts.shellPath = python_path;
+			opts.shellArgs = [script_path];
+			vscode.window.createTerminal(opts).show(false);;
```
Comment (apommel) [inline src/extension.ts:93]:
```suggestion
			vscode.window.createTerminal(opts).show(false);
```

**Comment by apommel** [comment]
Right, I checked one last time and everything seems fine, I'll merge it then. Thanks again!

**Pull request merged**